polar-core/src/rules.rs

```diff
@@ -16,24 +18,73 @@ impl Parameter {
     }
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+#[derive(Debug, Clone, Deserialize, Serialize)]
 pub struct Rule {
     pub name: Symbol,
     pub params: Vec<Parameter>,
     pub body: Term,
+    #[serde(skip, default = "SourceInfo::ffi")]
+    pub source_info: SourceInfo,
+}
+
+impl PartialEq for Rule {
+    fn eq(&self, other: &Self) -> bool {
+        self.name == other.name
+            && self.params.len() == other.params.len()
+            && self.params == other.params
+            && self.body == other.body
+    }
 }
 
 impl Rule {
     pub fn is_ground(&self) -> bool {
         self.params.iter().all(|p| p.is_ground())
     }
+
+    pub fn span(&self) -> Option<(usize, usize)> {
+        if let SourceInfo::Parser { left, right, .. } = self.source_info {
+            Some((left, right))
+        } else {
+            None
+        }
+    }
+
+    pub fn new_from_test(name: Symbol, params: Vec<Parameter>, body: Term) -> Self {
+        Self {
+            name,
+            params,
+            body,
+            source_info: SourceInfo::Test,
+        }
+    }
+
+    /// Creates a new term from the parser
+    pub fn new_from_parser(
+        src_id: u64,
+        left: usize,
+        right: usize,
+        name: Symbol,
+        params: Vec<Parameter>,
+        body: Term,
+    ) -> Self {
+        Self {
+            name,
+            params,
+            body,
+            source_info: SourceInfo::Parser {
+                src_id,
+                left,
+                right,
+            },
+        }
+    }
 }
 
 pub type Rules = Vec<Arc<Rule>>;
 
 type RuleSet = BTreeSet<u64>;
 
-#[derive(Clone, Default, Debug)]
+#[derive(Clone, Default, Debug, Deserialize, Serialize)]
```

Comment:
Why does this need to be serializable? It's an internal data structure to optimize looking up rules. Are we transferring the index and using it somewhere else?